## ECBM E6040 Homework 4 - Programming Problem

In [1]:
import matplotlib
import pylab
%matplotlib inline

from __future__ import print_function
import six.moves.cPickle as pickle
__docformat__ = 'restructedtext en'

from collections import OrderedDict
import copy
import numpy
import os
import random
import timeit

import theano
from theano import tensor as T

from hw4_utils import load_data, contextwin, shuffle, conlleval, check_dir
from hw4_nn import RNNSLU, myMLP, train_nn
#from hw4_utils import shared_dataset
from hw4a import test_rnnslu, RNNSLU2, test_rnnslu2
from hw4b import test_mlp_parity, RNN, test_rnn_parity

# Otherwise the deepcopy fails
import sys
sys.setrecursionlimit(1500)

Using gpu device 0: GRID K520


### Problem a: Recurrent Neural Network
#### Bullet 1: Experiment with RNNSLU

In [4]:
# TODO
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.59 best test F1 66.65
NEW BEST: epoch 1 valid F1 80.45 best test F1 77.65
NEW BEST: epoch 2 valid F1 83.67 best test F1 80.35
NEW BEST: epoch 3 valid F1 88.05 best test F1 84.08
NEW BEST: epoch 4 valid F1 89.42 best test F1 86.0
NEW BEST: epoch 5 valid F1 91.12 best test F1 86.69
NEW BEST: epoch 6 valid F1 91.55 best test F1 88.62
NEW BEST: epoch 7 valid F1 91.79 best test F1 89.73
NEW BEST: epoch 8 valid F1 93.63 best test F1 90.55
NEW BEST: epoch 9 valid F1 94.17 best test F1 91.56
NEW BEST: epoch 10 valid F1 94.45 best test F1 91.52
NEW BEST: epoch 11 valid F1 94.72 best test F1 92.45
NEW BEST: epoch 12 valid F1 95.31 best test F1 92.76
NEW BEST: epoch 13 valid F1 95.36 best test F1 92.87

NEW BEST: epoch 15 valid F1 95.64 best test F1

In [3]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 100,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 63.39 best test F1 60.51
NEW BEST: epoch 1 valid F1 75.46 best test F1 72.37
NEW BEST: epoch 2 valid F1 80.34 best test F1 77.06
NEW BEST: epoch 3 valid F1 83.89 best test F1 80.94
NEW BEST: epoch 4 valid F1 85.89 best test F1 82.83
NEW BEST: epoch 5 valid F1 88.02 best test F1 84.35
NEW BEST: epoch 6 valid F1 89.36 best test F1 85.14
NEW BEST: epoch 7 valid F1 90.96 best test F1 87.83
NEW BEST: epoch 8 valid F1 91.66 best test F1 88.68
NEW BEST: epoch 9 valid F1 93.03 best test F1 89.86
NEW BEST: epoch 10 valid F1 93.53 best test F1 90.73
NEW BEST: epoch 11 valid F1 94.0 best test F1 91.15
NEW BEST: epoch 12 valid F1 94.31 best test F1 91.34
NEW BEST: epoch 13 valid F1 94.62 best test F1 92.11
NEW BEST: epoch 14 valid F1 94.74 best test F1 

In [4]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 69.02 best test F1 66.48
NEW BEST: epoch 1 valid F1 79.95 best test F1 76.5
NEW BEST: epoch 2 valid F1 83.31 best test F1 80.87
NEW BEST: epoch 3 valid F1 87.5 best test F1 84.61
NEW BEST: epoch 4 valid F1 89.14 best test F1 85.84
NEW BEST: epoch 5 valid F1 90.91 best test F1 86.85
NEW BEST: epoch 6 valid F1 91.97 best test F1 88.43
NEW BEST: epoch 7 valid F1 92.96 best test F1 90.14
NEW BEST: epoch 8 valid F1 93.68 best test F1 90.61
NEW BEST: epoch 9 valid F1 94.03 best test F1 91.16
NEW BEST: epoch 10 valid F1 94.51 best test F1 91.04
NEW BEST: epoch 11 valid F1 94.59 best test F1 91.98
NEW BEST: epoch 12 valid F1 94.98 best test F1 92.46
NEW BEST: epoch 13 valid F1 95.31 best test F1 92.48
NEW BEST: epoch 14 valid F1 95.52 best test F1 9

In [5]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 100,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 62.15 best test F1 59.72
NEW BEST: epoch 1 valid F1 74.42 best test F1 71.73
NEW BEST: epoch 2 valid F1 80.26 best test F1 77.56
NEW BEST: epoch 3 valid F1 84.85 best test F1 81.3
NEW BEST: epoch 4 valid F1 86.43 best test F1 82.89
NEW BEST: epoch 5 valid F1 87.99 best test F1 83.89
NEW BEST: epoch 6 valid F1 89.4 best test F1 84.8
NEW BEST: epoch 7 valid F1 90.44 best test F1 86.72
NEW BEST: epoch 8 valid F1 91.27 best test F1 87.43
NEW BEST: epoch 9 valid F1 92.22 best test F1 88.28
NEW BEST: epoch 10 valid F1 92.5 best test F1 88.83
NEW BEST: epoch 11 valid F1 92.83 best test F1 89.58
NEW BEST: epoch 12 valid F1 93.25 best test F1 89.82
NEW BEST: epoch 13 valid F1 93.88 best test F1 90.93

NEW BEST: epoch 15 valid F1 94.75 best test F1 91

In [6]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: True
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.03 best test F1 64.56
NEW BEST: epoch 1 valid F1 79.49 best test F1 76.18
NEW BEST: epoch 2 valid F1 83.03 best test F1 80.44
NEW BEST: epoch 3 valid F1 87.15 best test F1 84.34
NEW BEST: epoch 4 valid F1 89.49 best test F1 85.76
NEW BEST: epoch 5 valid F1 91.09 best test F1 86.92
NEW BEST: epoch 6 valid F1 92.34 best test F1 88.95
NEW BEST: epoch 7 valid F1 93.49 best test F1 90.65
NEW BEST: epoch 8 valid F1 94.0 best test F1 91.13
NEW BEST: epoch 9 valid F1 94.46 best test F1 91.57
NEW BEST: epoch 10 valid F1 94.8 best test F1 91.14
NEW BEST: epoch 11 valid F1 95.49 best test F1 92.26
NEW BEST: epoch 12 valid F1 95.72 best test F1 92.42
NEW BEST: epoch 13 valid F1 95.83 best test F1 92.17
NEW BEST: epoch 14 valid F1 96.24 best test F1 

In [7]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 100,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: True
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 62.65 best test F1 59.55
NEW BEST: epoch 1 valid F1 75.05 best test F1 72.08
NEW BEST: epoch 2 valid F1 80.99 best test F1 77.46
NEW BEST: epoch 3 valid F1 84.95 best test F1 81.85
NEW BEST: epoch 4 valid F1 86.65 best test F1 84.34
NEW BEST: epoch 5 valid F1 88.46 best test F1 85.56
NEW BEST: epoch 6 valid F1 89.55 best test F1 86.67
NEW BEST: epoch 7 valid F1 91.23 best test F1 88.65
NEW BEST: epoch 8 valid F1 91.69 best test F1 89.37
NEW BEST: epoch 9 valid F1 92.89 best test F1 90.41
NEW BEST: epoch 10 valid F1 93.51 best test F1 90.75
NEW BEST: epoch 11 valid F1 94.32 best test F1 91.69


NEW BEST: epoch 14 valid F1 94.63 best test F1 92.8
NEW BEST: epoch 15 valid F1 95.29 best test F1 92.84
NEW BEST: epoch 16 valid F1 95.55 best test 

In [8]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: True
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.48 best test F1 64.68
NEW BEST: epoch 1 valid F1 79.43 best test F1 76.43
NEW BEST: epoch 2 valid F1 83.33 best test F1 80.35
NEW BEST: epoch 3 valid F1 86.98 best test F1 83.96
NEW BEST: epoch 4 valid F1 89.36 best test F1 85.8
NEW BEST: epoch 5 valid F1 90.63 best test F1 86.67
NEW BEST: epoch 6 valid F1 91.8 best test F1 88.71
NEW BEST: epoch 7 valid F1 92.49 best test F1 89.95
NEW BEST: epoch 8 valid F1 93.4 best test F1 90.83
NEW BEST: epoch 9 valid F1 93.78 best test F1 91.23
NEW BEST: epoch 10 valid F1 94.51 best test F1 91.32
NEW BEST: epoch 11 valid F1 94.73 best test F1 91.8
NEW BEST: epoch 12 valid F1 94.91 best test F1 92.5
NEW BEST: epoch 13 valid F1 95.38 best test F1 92.04

NEW BEST: epoch 15 valid F1 95.57 best test F1 93

In [9]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 100,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 60,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: True
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 63.09 best test F1 59.88
NEW BEST: epoch 1 valid F1 73.4 best test F1 70.95
NEW BEST: epoch 2 valid F1 79.93 best test F1 76.87
NEW BEST: epoch 3 valid F1 84.1 best test F1 80.57
NEW BEST: epoch 4 valid F1 86.25 best test F1 83.29
NEW BEST: epoch 5 valid F1 87.79 best test F1 84.72
NEW BEST: epoch 6 valid F1 89.0 best test F1 85.34
NEW BEST: epoch 7 valid F1 90.41 best test F1 88.11
NEW BEST: epoch 8 valid F1 91.73 best test F1 89.1
NEW BEST: epoch 9 valid F1 93.16 best test F1 90.46
NEW BEST: epoch 10 valid F1 93.36 best test F1 90.65
NEW BEST: epoch 11 valid F1 93.93 best test F1 91.35
NEW BEST: epoch 12 valid F1 94.61 best test F1 92.3

NEW BEST: epoch 14 valid F1 94.82 best test F1 92.5
NEW BEST: epoch 15 valid F1 95.34 best test F1 93.

nhidden: 200, win: 7, emb_dimension: 50, BEST RESULT: epoch 38 valid F1 97.16 best test F1 93.92 with the model


nhidden: 100, win: 7, emb_dimension: 50, BEST RESULT: epoch 59 valid F1 97.25 best test F1 93.89 with the model


nhidden: 200, win: 7, emb_dimension: 30, BEST RESULT: epoch 32 valid F1 97.1 best test F1 93.13 with the model


nhidden: 100, win: 7, emb_dimension: 30, BEST RESULT: epoch 47 valid F1 97.0 best test F1 93.62 with the model


nhidden: 200, win: 11, emb_dimension: 50, BEST RESULT: epoch 58 valid F1 97.32 best test F1 93.85 with the model


nhidden: 100, win: 11, emb_dimension: 50, BEST RESULT: epoch 57 valid F1 97.39 best test F1 93.92 with the model


nhidden: 200, win: 11, emb_dimension: 30, BEST RESULT: epoch 34 valid F1 97.07 best test F1 93.52 with the model


nhidden: 100, win: 11, emb_dimension: 30, BEST RESULT: epoch 59 valid F1 97.03 best test F1 93.87 with the model

Test accuracy has no obvious relation with the 3 parameters, nhidden, win and emb_dimension, since these parameters can all reach a relatively high accuracy. The parameters I used are not extreme enough to influence the accuracy distinctively.

#### Bullet 2: Experiment with RNNSLU without normalizing word embeddings

In [10]:
# TODO
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.03 best test F1 65.55
NEW BEST: epoch 1 valid F1 80.83 best test F1 77.94
NEW BEST: epoch 2 valid F1 83.8 best test F1 79.96
NEW BEST: epoch 3 valid F1 88.38 best test F1 84.19
NEW BEST: epoch 4 valid F1 89.69 best test F1 86.16
NEW BEST: epoch 5 valid F1 90.96 best test F1 86.9
NEW BEST: epoch 6 valid F1 91.34 best test F1 88.73
NEW BEST: epoch 7 valid F1 92.37 best test F1 90.21
NEW BEST: epoch 8 valid F1 93.89 best test F1 90.97
NEW BEST: epoch 9 valid F1 94.35 best test F1 91.3
NEW BEST: epoch 10 valid F1 94.78 best test F1 91.6

NEW BEST: epoch 12 valid F1 95.55 best test F1 92.93
NEW BEST: epoch 13 valid F1 95.62 best test F1 92.72


NEW BEST: epoch 16 valid F1 95.77 best test F1 93.26
NEW BEST: epoch 17 valid F1 96.07 best test F1

In [11]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 100,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 63.39 best test F1 60.51
NEW BEST: epoch 1 valid F1 75.46 best test F1 72.37
NEW BEST: epoch 2 valid F1 80.34 best test F1 77.06
NEW BEST: epoch 3 valid F1 83.89 best test F1 80.94
NEW BEST: epoch 4 valid F1 85.89 best test F1 82.83
NEW BEST: epoch 5 valid F1 88.02 best test F1 84.35
NEW BEST: epoch 6 valid F1 89.36 best test F1 85.14
NEW BEST: epoch 7 valid F1 90.96 best test F1 87.83
NEW BEST: epoch 8 valid F1 91.66 best test F1 88.68
NEW BEST: epoch 9 valid F1 93.03 best test F1 89.86
NEW BEST: epoch 10 valid F1 93.53 best test F1 90.73
NEW BEST: epoch 11 valid F1 94.0 best test F1 91.15
NEW BEST: epoch 12 valid F1 94.31 best test F1 91.34
NEW BEST: epoch 13 valid F1 94.62 best test F1 92.11
NEW BEST: epoch 14 valid F1 94.74 best test F1

In [12]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 69.02 best test F1 66.48
NEW BEST: epoch 1 valid F1 79.95 best test F1 76.5
NEW BEST: epoch 2 valid F1 83.31 best test F1 80.87
NEW BEST: epoch 3 valid F1 87.5 best test F1 84.61
NEW BEST: epoch 4 valid F1 89.14 best test F1 85.84
NEW BEST: epoch 5 valid F1 90.91 best test F1 86.85
NEW BEST: epoch 6 valid F1 91.97 best test F1 88.43
NEW BEST: epoch 7 valid F1 92.96 best test F1 90.14
NEW BEST: epoch 8 valid F1 93.68 best test F1 90.61
NEW BEST: epoch 9 valid F1 94.03 best test F1 91.16
NEW BEST: epoch 10 valid F1 94.51 best test F1 91.04
NEW BEST: epoch 11 valid F1 94.59 best test F1 91.98
NEW BEST: epoch 12 valid F1 94.98 best test F1 92.46
NEW BEST: epoch 13 valid F1 95.31 best test F1 92.48
NEW BEST: epoch 14 valid F1 95.52 best test F1 

In [3]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 100,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 62.15 best test F1 59.72
NEW BEST: epoch 1 valid F1 74.42 best test F1 71.73
NEW BEST: epoch 2 valid F1 80.26 best test F1 77.56
NEW BEST: epoch 3 valid F1 84.85 best test F1 81.3
NEW BEST: epoch 4 valid F1 86.43 best test F1 82.89
NEW BEST: epoch 5 valid F1 87.99 best test F1 83.89
NEW BEST: epoch 6 valid F1 89.4 best test F1 84.8
NEW BEST: epoch 7 valid F1 90.44 best test F1 86.72
NEW BEST: epoch 8 valid F1 91.27 best test F1 87.43
NEW BEST: epoch 9 valid F1 92.22 best test F1 88.28
NEW BEST: epoch 10 valid F1 92.5 best test F1 88.83
NEW BEST: epoch 11 valid F1 92.83 best test F1 89.58
NEW BEST: epoch 12 valid F1 93.25 best test F1 89.82
NEW BEST: epoch 13 valid F1 93.88 best test F1 90.93

NEW BEST: epoch 15 valid F1 94.75 best test F1 9

In [4]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: False
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.68 best test F1 65.09
NEW BEST: epoch 1 valid F1 79.2 best test F1 76.16
NEW BEST: epoch 2 valid F1 83.23 best test F1 80.23
NEW BEST: epoch 3 valid F1 87.24 best test F1 84.0
NEW BEST: epoch 4 valid F1 89.16 best test F1 85.52
NEW BEST: epoch 5 valid F1 91.31 best test F1 86.94
NEW BEST: epoch 6 valid F1 91.94 best test F1 88.44
NEW BEST: epoch 7 valid F1 93.64 best test F1 90.68
NEW BEST: epoch 8 valid F1 94.0 best test F1 90.7
NEW BEST: epoch 9 valid F1 94.7 best test F1 91.84
NEW BEST: epoch 10 valid F1 95.26 best test F1 91.01
NEW BEST: epoch 11 valid F1 95.38 best test F1 92.38
NEW BEST: epoch 12 valid F1 95.66 best test F1 92.6
NEW BEST: epoch 13 valid F1 95.81 best test F1 92.41
NEW BEST: epoch 14 valid F1 96.07 best test F1 92.

In [5]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 100,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: False
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 61.86 best test F1 58.41
NEW BEST: epoch 1 valid F1 74.97 best test F1 71.49
NEW BEST: epoch 2 valid F1 81.15 best test F1 77.79
NEW BEST: epoch 3 valid F1 84.83 best test F1 82.47
NEW BEST: epoch 4 valid F1 86.79 best test F1 84.02
NEW BEST: epoch 5 valid F1 88.1 best test F1 84.62
NEW BEST: epoch 6 valid F1 89.72 best test F1 86.07
NEW BEST: epoch 7 valid F1 91.0 best test F1 88.43
NEW BEST: epoch 8 valid F1 91.41 best test F1 88.58
NEW BEST: epoch 9 valid F1 92.93 best test F1 90.56
NEW BEST: epoch 10 valid F1 93.48 best test F1 90.85
NEW BEST: epoch 11 valid F1 93.72 best test F1 91.39
NEW BEST: epoch 12 valid F1 94.45 best test F1 92.32
NEW BEST: epoch 13 valid F1 94.84 best test F1 91.91

NEW BEST: epoch 15 valid F1 95.46 best test F

In [6]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: False
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.48 best test F1 64.68
NEW BEST: epoch 1 valid F1 79.43 best test F1 76.43
NEW BEST: epoch 2 valid F1 83.33 best test F1 80.35
NEW BEST: epoch 3 valid F1 86.98 best test F1 83.96
NEW BEST: epoch 4 valid F1 89.36 best test F1 85.8
NEW BEST: epoch 5 valid F1 90.63 best test F1 86.67
NEW BEST: epoch 6 valid F1 91.8 best test F1 88.71
NEW BEST: epoch 7 valid F1 92.49 best test F1 89.95
NEW BEST: epoch 8 valid F1 93.4 best test F1 90.83
NEW BEST: epoch 9 valid F1 93.78 best test F1 91.23
NEW BEST: epoch 10 valid F1 94.51 best test F1 91.32
NEW BEST: epoch 11 valid F1 94.73 best test F1 91.8
NEW BEST: epoch 12 valid F1 94.91 best test F1 92.5
NEW BEST: epoch 13 valid F1 95.38 best test F1 92.04

NEW BEST: epoch 15 valid F1 95.57 best test F1 9

In [7]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden': 100,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 60,
    'savemodel': False,
    'normal': False,
    'folder':'../result'}
test_rnnslu(**param)

verbose: True
normal: False
win: 11
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 62.65 best test F1 59.08
NEW BEST: epoch 1 valid F1 73.53 best test F1 70.88
NEW BEST: epoch 2 valid F1 79.89 best test F1 76.64
NEW BEST: epoch 3 valid F1 84.25 best test F1 80.59
NEW BEST: epoch 4 valid F1 86.46 best test F1 83.46
NEW BEST: epoch 5 valid F1 87.76 best test F1 84.62
NEW BEST: epoch 6 valid F1 88.81 best test F1 85.43
NEW BEST: epoch 7 valid F1 90.47 best test F1 88.21
NEW BEST: epoch 8 valid F1 91.57 best test F1 88.99
NEW BEST: epoch 9 valid F1 93.17 best test F1 90.56

NEW BEST: epoch 11 valid F1 93.68 best test F1 91.62
NEW BEST: epoch 12 valid F1 94.42 best test F1 92.02
NEW BEST: epoch 13 valid F1 94.66 best test F1 91.82
NEW BEST: epoch 14 valid F1 94.75 best test F1 92.29
NEW BEST: epoch 15 valid F1 95.54 best test

normal: False,


nhidden: 200, win: 7, emb_dimension: 50, BEST RESULT: epoch 57 valid F1 97.29 best test F1 94.28 with the model


nhidden: 100, win: 7, emb_dimension: 50, BEST RESULT: epoch 59 valid F1 97.25 best test F1 93.89 with the model


nhidden: 200, win: 7, emb_dimension: 30, BEST RESULT: epoch 32 valid F1 97.1 best test F1 93.13 with the model


nhidden: 100, win: 7, emb_dimension: 30, BEST RESULT: epoch 47 valid F1 97.0 best test F1 93.62 with the model


nhidden: 200, win: 11, emb_dimension: 50, BEST RESULT: epoch 57 valid F1 97.22 best test F1 93.78 with the model


nhidden: 100, win: 11, emb_dimension: 50, BEST RESULT: epoch 52 valid F1 97.11 best test F1 93.63 with the model


nhidden: 200, win: 11, emb_dimension: 30, BEST RESULT: epoch 34 valid F1 97.07 best test F1 93.52 with the model


nhidden: 100, win: 11, emb_dimension: 30, BEST RESULT: epoch 38 valid F1 96.85 best test F1 93.74 with the model


Test accuracy has no obvious relation with the 4 parameters, nhidden, win, emb_dimension and normal, since these parameters can all reach a relatively high accuracy. The parameters I used are not extreme enough to influence the accuracy distinctively.

#### Bullet 3: Implement RNN with two hidden layers

In [2]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden1': 100,
    'nhidden2': 100,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 100
win: 7
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 54.83 best test F1 50.99
NEW BEST: epoch 1 valid F1 66.41 best test F1 63.77
NEW BEST: epoch 2 valid F1 72.79 best test F1 71.05
NEW BEST: epoch 3 valid F1 78.32 best test F1 73.89
NEW BEST: epoch 4 valid F1 82.03 best test F1 78.58
NEW BEST: epoch 5 valid F1 84.2 best test F1 81.81
NEW BEST: epoch 6 valid F1 85.75 best test F1 82.97
NEW BEST: epoch 7 valid F1 88.78 best test F1 85.65
NEW BEST: epoch 8 valid F1 89.0 best test F1 86.31
NEW BEST: epoch 9 valid F1 89.44 best test F1 86.19
NEW BEST: epoch 10 valid F1 90.5 best test F1 87.15
NEW BEST: epoch 11 valid F1 91.56 best test F1 88.19
NEW BEST: epoch 12 valid F1 92.33 best test F1 88.6


NEW BEST: epoch 15 valid F1 93.13 best test F1 89.71

NEW BEST: epoch 17 valid F1 93.3

In [3]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden1': 150,
    'nhidden2': 150,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 150
win: 7
nhidden1: 150
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 58.04 best test F1 53.57
NEW BEST: epoch 1 valid F1 69.51 best test F1 69.25
NEW BEST: epoch 2 valid F1 77.27 best test F1 75.51
NEW BEST: epoch 3 valid F1 81.52 best test F1 78.55
NEW BEST: epoch 4 valid F1 83.91 best test F1 80.8
NEW BEST: epoch 5 valid F1 87.49 best test F1 83.59
NEW BEST: epoch 6 valid F1 88.32 best test F1 84.49
NEW BEST: epoch 7 valid F1 89.3 best test F1 87.69
NEW BEST: epoch 8 valid F1 91.01 best test F1 87.96
NEW BEST: epoch 9 valid F1 91.42 best test F1 88.1
NEW BEST: epoch 10 valid F1 91.82 best test F1 88.76
NEW BEST: epoch 11 valid F1 92.43 best test F1 89.48
NEW BEST: epoch 12 valid F1 93.34 best test F1 89.74
NEW BEST: epoch 13 valid F1 93.96 best test F1 91.07

NEW BEST: epoch 15 valid F1 94.01

In [4]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden1': 150,
    'nhidden2': 100,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 100
win: 7
nhidden1: 150
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.33 best test F1 51.27
NEW BEST: epoch 1 valid F1 66.57 best test F1 63.26
NEW BEST: epoch 2 valid F1 70.54 best test F1 68.59
NEW BEST: epoch 3 valid F1 78.54 best test F1 75.62
NEW BEST: epoch 4 valid F1 82.25 best test F1 79.67
NEW BEST: epoch 5 valid F1 84.64 best test F1 81.95
NEW BEST: epoch 6 valid F1 85.64 best test F1 83.18
NEW BEST: epoch 7 valid F1 87.57 best test F1 84.51
NEW BEST: epoch 8 valid F1 89.0 best test F1 86.76
NEW BEST: epoch 9 valid F1 89.41 best test F1 86.69
NEW BEST: epoch 10 valid F1 90.43 best test F1 87.15
NEW BEST: epoch 11 valid F1 91.13 best test F1 88.15
NEW BEST: epoch 12 valid F1 91.96 best test F1 88.53
NEW BEST: epoch 13 valid F1 92.73 best test F1 89.42

NEW BEST: epoch 15 valid F1 93.

In [5]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden1': 100,
    'nhidden2': 150,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 150
win: 7
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.11 best test F1 52.01
NEW BEST: epoch 1 valid F1 66.29 best test F1 63.06
NEW BEST: epoch 2 valid F1 73.54 best test F1 70.51
NEW BEST: epoch 3 valid F1 79.56 best test F1 76.99
NEW BEST: epoch 4 valid F1 83.79 best test F1 80.7
NEW BEST: epoch 5 valid F1 85.31 best test F1 82.45
NEW BEST: epoch 6 valid F1 86.87 best test F1 83.51
NEW BEST: epoch 7 valid F1 89.25 best test F1 86.73
NEW BEST: epoch 8 valid F1 89.95 best test F1 87.02
NEW BEST: epoch 9 valid F1 90.8 best test F1 87.31
NEW BEST: epoch 10 valid F1 91.42 best test F1 87.98
NEW BEST: epoch 11 valid F1 91.7 best test F1 89.16
NEW BEST: epoch 12 valid F1 92.47 best test F1 89.7
NEW BEST: epoch 13 valid F1 93.34 best test F1 89.34

NEW BEST: epoch 15 valid F1 93.96 

In [2]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden1': 100,
    'nhidden2': 100,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 100
win: 11
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 55.5 best test F1 50.97
NEW BEST: epoch 1 valid F1 64.98 best test F1 62.28
NEW BEST: epoch 2 valid F1 72.08 best test F1 70.09
NEW BEST: epoch 3 valid F1 78.42 best test F1 76.29
NEW BEST: epoch 4 valid F1 82.8 best test F1 79.53
NEW BEST: epoch 5 valid F1 84.56 best test F1 81.48
NEW BEST: epoch 6 valid F1 86.26 best test F1 82.53
NEW BEST: epoch 7 valid F1 88.04 best test F1 85.84
NEW BEST: epoch 8 valid F1 88.71 best test F1 86.42
NEW BEST: epoch 9 valid F1 89.91 best test F1 86.53
NEW BEST: epoch 10 valid F1 90.7 best test F1 87.35
NEW BEST: epoch 11 valid F1 91.26 best test F1 87.92
NEW BEST: epoch 12 valid F1 92.08 best test F1 89.17
NEW BEST: epoch 13 valid F1 92.52 best test F1 88.86


NEW BEST: epoch 16 valid F1 93.

In [3]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden1': 150,
    'nhidden2': 150,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 150
win: 11
nhidden1: 150
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 58.49 best test F1 53.8
NEW BEST: epoch 1 valid F1 69.97 best test F1 67.48
NEW BEST: epoch 2 valid F1 78.04 best test F1 74.95
NEW BEST: epoch 3 valid F1 82.03 best test F1 78.57
NEW BEST: epoch 4 valid F1 84.7 best test F1 80.92
NEW BEST: epoch 5 valid F1 86.57 best test F1 83.15
NEW BEST: epoch 6 valid F1 87.44 best test F1 84.77
NEW BEST: epoch 7 valid F1 89.75 best test F1 86.91
NEW BEST: epoch 8 valid F1 90.2 best test F1 87.22
NEW BEST: epoch 9 valid F1 91.0 best test F1 87.53
NEW BEST: epoch 10 valid F1 91.12 best test F1 87.18
NEW BEST: epoch 11 valid F1 92.39 best test F1 88.31
NEW BEST: epoch 12 valid F1 92.65 best test F1 89.08
NEW BEST: epoch 13 valid F1 93.21 best test F1 89.6

NEW BEST: epoch 15 valid F1 93.73 

In [4]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden1': 150,
    'nhidden2': 100,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 100
win: 11
nhidden1: 150
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 55.98 best test F1 51.79
NEW BEST: epoch 1 valid F1 67.7 best test F1 64.6
NEW BEST: epoch 2 valid F1 75.71 best test F1 72.5
NEW BEST: epoch 3 valid F1 79.43 best test F1 76.16
NEW BEST: epoch 4 valid F1 82.39 best test F1 79.56
NEW BEST: epoch 5 valid F1 84.85 best test F1 81.68
NEW BEST: epoch 6 valid F1 87.05 best test F1 82.61
NEW BEST: epoch 7 valid F1 88.36 best test F1 85.65
NEW BEST: epoch 8 valid F1 88.56 best test F1 87.17
NEW BEST: epoch 9 valid F1 89.94 best test F1 87.78
NEW BEST: epoch 10 valid F1 91.03 best test F1 88.01

NEW BEST: epoch 12 valid F1 93.37 best test F1 89.86


NEW BEST: epoch 15 valid F1 94.03 best test F1 90.83

NEW BEST: epoch 17 valid F1 94.73 best test F1 91.46
NEW BEST: epoch 18 valid F1 9

In [5]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden1': 100,
    'nhidden2': 150,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 150
win: 11
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.86 best test F1 53.27
NEW BEST: epoch 1 valid F1 67.74 best test F1 65.52
NEW BEST: epoch 2 valid F1 75.44 best test F1 73.67
NEW BEST: epoch 3 valid F1 79.99 best test F1 77.6
NEW BEST: epoch 4 valid F1 83.86 best test F1 81.16
NEW BEST: epoch 5 valid F1 85.03 best test F1 82.86
NEW BEST: epoch 6 valid F1 86.56 best test F1 84.42
NEW BEST: epoch 7 valid F1 88.65 best test F1 86.92
NEW BEST: epoch 8 valid F1 89.47 best test F1 86.96
NEW BEST: epoch 9 valid F1 90.21 best test F1 87.06
NEW BEST: epoch 10 valid F1 91.46 best test F1 87.94
NEW BEST: epoch 11 valid F1 92.07 best test F1 88.77
NEW BEST: epoch 12 valid F1 92.8 best test F1 88.92
NEW BEST: epoch 13 valid F1 93.62 best test F1 90.87
NEW BEST: epoch 14 valid F1 93.6

In [ ]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden1': 100,
    'nhidden2': 100,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 100
win: 7
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.49 best test F1 52.42
NEW BEST: epoch 1 valid F1 65.16 best test F1 62.25
NEW BEST: epoch 2 valid F1 74.16 best test F1 71.63
NEW BEST: epoch 3 valid F1 78.55 best test F1 75.6
NEW BEST: epoch 4 valid F1 82.42 best test F1 79.36
NEW BEST: epoch 5 valid F1 83.93 best test F1 80.9
NEW BEST: epoch 6 valid F1 85.6 best test F1 82.0
NEW BEST: epoch 7 valid F1 87.31 best test F1 84.85
NEW BEST: epoch 8 valid F1 87.95 best test F1 86.05
NEW BEST: epoch 9 valid F1 89.14 best test F1 85.78
NEW BEST: epoch 10 valid F1 89.63 best test F1 86.72
NEW BEST: epoch 11 valid F1 90.09 best test F1 87.09
NEW BEST: epoch 12 valid F1 91.42 best test F1 88.89
NEW BEST: epoch 13 valid F1 91.6 best test F1 88.96
NEW BEST: epoch 14 valid F1 91.87 be

In [ ]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden1': 150,
    'nhidden2': 150,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 150
win: 7
nhidden1: 150
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 58.92 best test F1 55.77
NEW BEST: epoch 1 valid F1 70.27 best test F1 67.57
NEW BEST: epoch 2 valid F1 76.76 best test F1 75.51
NEW BEST: epoch 3 valid F1 81.68 best test F1 79.49
NEW BEST: epoch 4 valid F1 84.93 best test F1 82.24
NEW BEST: epoch 5 valid F1 86.6 best test F1 82.97
NEW BEST: epoch 6 valid F1 88.64 best test F1 85.01
NEW BEST: epoch 7 valid F1 89.65 best test F1 86.95
NEW BEST: epoch 8 valid F1 90.9 best test F1 87.99
NEW BEST: epoch 9 valid F1 91.08 best test F1 88.02
NEW BEST: epoch 10 valid F1 91.61 best test F1 88.8
NEW BEST: epoch 11 valid F1 91.98 best test F1 89.64
NEW BEST: epoch 12 valid F1 93.03 best test F1 89.86
NEW BEST: epoch 13 valid F1 93.38 best test F1 90.8
NEW BEST: epoch 14 valid F1 93.69 b

In [ ]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden1': 150,
    'nhidden2': 100,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 100
win: 7
nhidden1: 150
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 57.32 best test F1 53.83
NEW BEST: epoch 1 valid F1 66.77 best test F1 66.17
NEW BEST: epoch 2 valid F1 74.73 best test F1 73.34
NEW BEST: epoch 3 valid F1 80.35 best test F1 77.91
NEW BEST: epoch 4 valid F1 84.27 best test F1 81.63
NEW BEST: epoch 5 valid F1 85.37 best test F1 82.17
NEW BEST: epoch 6 valid F1 87.35 best test F1 83.9
NEW BEST: epoch 7 valid F1 88.65 best test F1 85.19
NEW BEST: epoch 8 valid F1 89.35 best test F1 86.54
NEW BEST: epoch 9 valid F1 90.01 best test F1 86.82
NEW BEST: epoch 10 valid F1 90.91 best test F1 87.78
NEW BEST: epoch 11 valid F1 91.18 best test F1 88.39
NEW BEST: epoch 12 valid F1 92.39 best test F1 89.95
NEW BEST: epoch 13 valid F1 92.71 best test F1 90.69


NEW BEST: epoch 16 valid F1 93

In [5]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden1': 100,
    'nhidden2': 150,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 150
win: 7
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 57.02 best test F1 54.02
NEW BEST: epoch 1 valid F1 68.56 best test F1 66.54
NEW BEST: epoch 2 valid F1 74.65 best test F1 73.28
NEW BEST: epoch 3 valid F1 79.28 best test F1 76.95
NEW BEST: epoch 4 valid F1 83.06 best test F1 81.17
NEW BEST: epoch 5 valid F1 84.56 best test F1 82.51
NEW BEST: epoch 6 valid F1 87.26 best test F1 82.84
NEW BEST: epoch 7 valid F1 88.58 best test F1 85.15
NEW BEST: epoch 8 valid F1 89.36 best test F1 86.95
NEW BEST: epoch 9 valid F1 89.66 best test F1 86.69
NEW BEST: epoch 10 valid F1 90.89 best test F1 87.96
NEW BEST: epoch 11 valid F1 91.06 best test F1 88.21
NEW BEST: epoch 12 valid F1 91.77 best test F1 88.44
NEW BEST: epoch 13 valid F1 92.26 best test F1 89.79

NEW BEST: epoch 15 valid F1 92

In [6]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden1': 100,
    'nhidden2': 100,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 100
win: 11
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.38 best test F1 52.89
NEW BEST: epoch 1 valid F1 66.09 best test F1 62.83
NEW BEST: epoch 2 valid F1 74.74 best test F1 73.01
NEW BEST: epoch 3 valid F1 80.46 best test F1 77.22
NEW BEST: epoch 4 valid F1 83.31 best test F1 80.52
NEW BEST: epoch 5 valid F1 84.93 best test F1 81.67
NEW BEST: epoch 6 valid F1 86.39 best test F1 83.15
NEW BEST: epoch 7 valid F1 88.06 best test F1 84.28

NEW BEST: epoch 9 valid F1 88.25 best test F1 85.15
NEW BEST: epoch 10 valid F1 89.86 best test F1 86.89
NEW BEST: epoch 11 valid F1 89.97 best test F1 87.92
NEW BEST: epoch 12 valid F1 91.06 best test F1 88.31
NEW BEST: epoch 13 valid F1 91.88 best test F1 89.45
NEW BEST: epoch 14 valid F1 92.23 best test F1 88.35

NEW BEST: epoch 16 valid F1

In [7]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden1': 150,
    'nhidden2': 150,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 150
win: 11
nhidden1: 150
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 58.91 best test F1 54.81
NEW BEST: epoch 1 valid F1 69.0 best test F1 67.04
NEW BEST: epoch 2 valid F1 76.92 best test F1 75.19
NEW BEST: epoch 3 valid F1 82.08 best test F1 79.24
NEW BEST: epoch 4 valid F1 84.38 best test F1 82.15
NEW BEST: epoch 5 valid F1 85.96 best test F1 83.13
NEW BEST: epoch 6 valid F1 88.25 best test F1 85.38
NEW BEST: epoch 7 valid F1 89.49 best test F1 86.83
NEW BEST: epoch 8 valid F1 90.75 best test F1 88.99
NEW BEST: epoch 9 valid F1 91.05 best test F1 88.44
NEW BEST: epoch 10 valid F1 91.54 best test F1 89.21
NEW BEST: epoch 11 valid F1 92.35 best test F1 90.45
NEW BEST: epoch 12 valid F1 92.89 best test F1 90.61
NEW BEST: epoch 13 valid F1 93.48 best test F1 90.76


NEW BEST: epoch 16 valid F1 9

In [8]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden1': 150,
    'nhidden2': 100,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 100
win: 11
nhidden1: 150
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.61 best test F1 53.01
NEW BEST: epoch 1 valid F1 67.29 best test F1 65.04
NEW BEST: epoch 2 valid F1 74.23 best test F1 72.67
NEW BEST: epoch 3 valid F1 80.4 best test F1 77.02
NEW BEST: epoch 4 valid F1 83.82 best test F1 80.89
NEW BEST: epoch 5 valid F1 84.65 best test F1 81.41
NEW BEST: epoch 6 valid F1 86.93 best test F1 83.46
NEW BEST: epoch 7 valid F1 88.23 best test F1 85.24
NEW BEST: epoch 8 valid F1 89.02 best test F1 86.78
NEW BEST: epoch 9 valid F1 89.98 best test F1 86.78
NEW BEST: epoch 10 valid F1 90.42 best test F1 87.09
NEW BEST: epoch 11 valid F1 90.95 best test F1 88.12
NEW BEST: epoch 12 valid F1 91.92 best test F1 88.5
NEW BEST: epoch 13 valid F1 92.49 best test F1 89.15

NEW BEST: epoch 15 valid F1 93.

In [ ]:
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 11,
    'nhidden1': 100,
    'nhidden2': 150,
    'seed': 345,
    'emb_dimension': 30,
    'nepochs': 50,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
test_rnnslu2(**param)

verbose: True
normal: True
nhidden2: 150
win: 11
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 30
nepochs: 50
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.24 best test F1 52.23
NEW BEST: epoch 1 valid F1 67.74 best test F1 64.68
NEW BEST: epoch 2 valid F1 74.37 best test F1 71.57
NEW BEST: epoch 3 valid F1 80.46 best test F1 76.72
NEW BEST: epoch 4 valid F1 83.29 best test F1 80.06
NEW BEST: epoch 5 valid F1 86.13 best test F1 81.72
NEW BEST: epoch 6 valid F1 86.28 best test F1 83.42
NEW BEST: epoch 7 valid F1 88.84 best test F1 86.32
NEW BEST: epoch 8 valid F1 89.66 best test F1 87.33
NEW BEST: epoch 9 valid F1 90.32 best test F1 87.45
NEW BEST: epoch 10 valid F1 91.09 best test F1 88.03
NEW BEST: epoch 11 valid F1 91.35 best test F1 88.64
NEW BEST: epoch 12 valid F1 92.07 best test F1 89.17
NEW BEST: epoch 13 valid F1 93.29 best test F1 90.2



NEW BEST: epoch 17 valid F1 

'win': 7, 'nhidden1': 100, 'nhidden2': 100, 'emb_dimension': 50, BEST RESULT: epoch 47 valid F1 95.89 best test F1 92.27 with the model

'win': 7, 'nhidden1': 150, 'nhidden2': 150, 'emb_dimension': 50, BEST RESULT: epoch 47 valid F1 96.43 best test F1 93.41 with the model

'win': 7, 'nhidden1': 150, 'nhidden2': 100, 'emb_dimension': 50, BEST RESULT: epoch 39 valid F1 96.19 best test F1 92.51 with the model

'win': 7, 'nhidden1': 100, 'nhidden2': 150, 'emb_dimension': 50, BEST RESULT: epoch 47 valid F1 95.87 best test F1 92.49 with the model

'win': 11, 'nhidden1': 100, 'nhidden2': 100, 'emb_dimension': 50, BEST RESULT: epoch 47 valid F1 95.95 best test F1 92.67 with the model

'win': 11, 'nhidden1': 150, 'nhidden2': 150, 'emb_dimension': 50, BEST RESULT: epoch 47 valid F1 96.02 best test F1 92.33 with the model

'win': 11, 'nhidden1': 150, 'nhidden2': 100, 'emb_dimension': 50, BEST RESULT: epoch 47 valid F1 96.42 best test F1 92.52 with the model

'win': 11, 'nhidden1': 100, 'nhidden2': 150, 'emb_dimension': 50, BEST RESULT: epoch 34 valid F1 96.48 best test F1 91.84 with the model

'win': 7, 'nhidden1': 100, 'nhidden2': 100, 'emb_dimension': 30, BEST RESULT: epoch 49 valid F1 96.02 best test F1 92.43 with the model

'win': 7, 'nhidden1': 150, 'nhidden2': 150, 'emb_dimension': 30, BEST RESULT: epoch 46 valid F1 96.26 best test F1 92.56 with the model

'win': 7, 'nhidden1': 150, 'nhidden2': 100, 'emb_dimension': 30, BEST RESULT: epoch 47 valid F1 96.26 best test F1 93.09 with the model

'win': 7, 'nhidden1': 100, 'nhidden2': 150, 'emb_dimension': 30, BEST RESULT: epoch 47 valid F1 96.05 best test F1 92.38 with the model

'win': 11, 'nhidden1': 100, 'nhidden2': 100, 'emb_dimension': 30, BEST RESULT: epoch 48 valid F1 95.85 best test F1 92.11 with the model

'win': 11, 'nhidden1': 150, 'nhidden2': 150, 'emb_dimension': 30, BEST RESULT: epoch 48 valid F1 96.0 best test F1 92.46 with the model

'win': 11, 'nhidden1': 150, 'nhidden2': 100, 'emb_dimension': 30, BEST RESULT: epoch 48 valid F1 95.95 best test F1 92.32 with the model

'win': 11, 'nhidden1': 100, 'nhidden2': 150, 'emb_dimension': 30, BEST RESULT: epoch 48 valid F1 95.77 best test F1 91.66 with the model


RNN wiht two hidden layers also has a high accurancy. Test accuracy has no obvious relation with the 4 parameters, nhidden1, nhidden2, win and emb_dimension, since these parameters can all reach a relatively high accuracy. The parameters I used are not extreme enough to influence the accuracy distinctively. But the average accuracy of 2 hidden layers is slightly lower than that of 1 hidden layer.

#### Bullet 4: Experiment with word embeddings

In [5]:
# TODO
param = {
    'fold': 3,
    'lr': 0.0970806646812754,
    'verbose': True,
    'decay': True,
    'win': 7,
    'nhidden': 200,
    'seed': 345,
    'emb_dimension': 50,
    'nepochs': 20,
    'savemodel': False,
    'normal': True,
    'folder':'../result'}
[emb, dic] = test_rnnslu(**param)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 20
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.59 best test F1 66.65
NEW BEST: epoch 1 valid F1 80.45 best test F1 77.65
NEW BEST: epoch 2 valid F1 83.67 best test F1 80.35
NEW BEST: epoch 3 valid F1 88.05 best test F1 84.08
NEW BEST: epoch 4 valid F1 89.42 best test F1 86.0
NEW BEST: epoch 5 valid F1 91.12 best test F1 86.69
NEW BEST: epoch 6 valid F1 91.55 best test F1 88.62
NEW BEST: epoch 7 valid F1 91.79 best test F1 89.73
NEW BEST: epoch 8 valid F1 93.63 best test F1 90.55
NEW BEST: epoch 9 valid F1 94.17 best test F1 91.56
NEW BEST: epoch 10 valid F1 94.45 best test F1 91.52
NEW BEST: epoch 11 valid F1 94.72 best test F1 92.45
NEW BEST: epoch 12 valid F1 95.31 best test F1 92.76
NEW BEST: epoch 13 valid F1 95.36 best test F1 92.87

NEW BEST: epoch 15 valid F1 95.64 best test F1

In [23]:
print(dic['words2idx'])

{'all': 32, 'coach': 110, 'cincinnati': 102, 'month': 318, 'four': 202, 'code': 111, 'go': 213, 'departing': 139, 'thursday': 496, 'to': 502, 'dinnertime': 151, 'under': 529, 'sorry': 450, 'include': 235, 'midwest': 311, 'worth': 564, 'southwest': 451, 'me': 301, 'returning': 408, 'far': 181, 'vegas': 539, 'airfare': 24, 'ticket': 498, 'difference': 148, 'arrange': 54, 'tickets': 499, 'louis': 286, 'cheapest': 99, 'list': 276, 'wednesday': 546, 'leave': 268, 'heading': 222, 'ten': 474, 'direct': 152, 'turboprop': 520, 'rate': 395, 'cost': 121, 'quebec': 392, 'layover': 266, 'choices': 101, 'what': 554, 'stands': 454, 'chicago': 100, 'reaching': 397, 'transcontinental': 510, 'goes': 214, 'new': 332, 'transportation': 512, 'fifteenth': 187, 'hours': 228, 'let': 272, 'twentieth': 523, 'along': 33, 'thrift': 494, 'passengers': 371, 'great': 216, 'thirty': 490, 'canadian': 91, 'leaves': 269, 'alaska': 31, 'leaving': 270, 'amount': 38, 'weekday': 549, 'makes': 294, 'midway': 310, 'montreal':

In [128]:
print(dic['words2idx']['afternoon'])

20


In [129]:
print(dic['words2idx']['pm'])

384


In [130]:
print(dic['words2idx']['morning'])

321


In [131]:
print(dic['words2idx']['am'])

35


In [133]:
print(numpy.linalg.norm(emb.get_value(borrow=True)[20]-emb.get_value(borrow=True)[384]))

0.96103


In [134]:
print(numpy.linalg.norm(emb.get_value(borrow=True)[321]-emb.get_value(borrow=True)[35]))

1.04325


In [132]:
a = emb.get_value(borrow=True)[20]-emb.get_value(borrow=True)[384]
b = emb.get_value(borrow=True)[321]-emb.get_value(borrow=True)[35]
print(numpy.linalg.norm(a - b))

0.953251


In [137]:
print(emb.get_value(borrow=True)[20]-emb.get_value(borrow=True)[384])

[ 0.08398601  0.15741269 -0.30875421  0.08295178 -0.24993479 -0.20839174
 -0.03143764  0.12233616  0.06618224  0.16798201 -0.00245194 -0.11969084
  0.21668237  0.03886913  0.13738866 -0.24387436 -0.04371469 -0.20264718
 -0.05424038  0.06691113  0.02992406  0.10118935 -0.02943821  0.03427172
  0.03152721 -0.03780183 -0.18142615  0.04435935 -0.15862909 -0.06338631
 -0.05471008  0.22534055 -0.14117867 -0.04643361 -0.21017429  0.01771721
 -0.01124765  0.11808204 -0.09267631 -0.01287176  0.12290359 -0.17545927
 -0.11464086 -0.04628494  0.16033921  0.06636767 -0.17285439  0.03080034
  0.24492861  0.19649224]


In [136]:
print(emb.get_value(borrow=True)[321]-emb.get_value(borrow=True)[35])

[ 0.10467197  0.06095466 -0.32516316  0.12160608  0.01035052 -0.24425858
  0.0567383  -0.01250575 -0.08429492  0.11474668  0.01051248 -0.0968879
  0.14674775 -0.0532364   0.06008875 -0.06433021 -0.28264406 -0.34927326
  0.04970416 -0.02992047  0.06664354  0.20487639 -0.00719279 -0.05557647
  0.03290561  0.067352   -0.03335024 -0.04922209 -0.08860435  0.10275987
  0.169218    0.24765089 -0.0260942  -0.2758404   0.06014173 -0.07231296
  0.29180974 -0.11329182  0.08229014 -0.00553752  0.15784326  0.10791753
 -0.1746684  -0.23649044  0.06000485  0.01566779 -0.2024267  -0.01871086
  0.13327892  0.28190523]


I choose 'morning' and 'am', 'afternoon' and 'pm', and I can see that F("morning") - F("am") is almost equal to  F("afternoon") - F("pm") by 

### Problem b: Learning Parity Function
#### Bullet 1: Shallow MLP

In [11]:
# 8 bits
accuracy = test_mlp_parity(learning_rate=1, L1_reg=0.00, L2_reg=0.00, n_epochs=1000,
             batch_size=4, n_hidden=8000, n_hiddenLayers=1,
             verbose=False, n_bit = 8)

... building the model
... training
Optimization complete.
Best validation score of 0.000000 % obtained at iteration 3750, with test performance 0.000000 %


The training process for function test_mlp_parity ran for 0.18m


In [19]:
# 12 bits
accuracy = test_mlp_parity(learning_rate=1, L1_reg=0.00, L2_reg=0.00, n_epochs=15000,
             batch_size=8, n_hidden=20000, n_hiddenLayers=1,
             verbose=False, n_bit = 12)

... building the model
... training
Optimization complete.
Best validation score of 0.000000 % obtained at iteration 8500, with test performance 0.000000 %


The training process for function test_mlp_parity ran for 0.65m


#### Bullet 2: Deep MLP

In [18]:
accuracy = test_mlp_parity(learning_rate=0.1, L1_reg=0.00, L2_reg=0.00, n_epochs=1000,
             batch_size=50, n_hidden=200, n_hiddenLayers=5,
             verbose=True, n_bit = 8)

... building the model
... training
training @ iter =  0
epoch 1, minibatch 20/20, validation error 48.400000 %
     epoch 1, minibatch 20/20, test error of best model 55.000000 %
epoch 2, minibatch 20/20, validation error 46.200000 %
     epoch 2, minibatch 20/20, test error of best model 52.000000 %
epoch 3, minibatch 20/20, validation error 43.000000 %
     epoch 3, minibatch 20/20, test error of best model 50.000000 %
epoch 4, minibatch 20/20, validation error 45.800000 %
epoch 5, minibatch 20/20, validation error 45.400000 %
training @ iter =  100
epoch 6, minibatch 20/20, validation error 46.800000 %
epoch 7, minibatch 20/20, validation error 45.800000 %
epoch 8, minibatch 20/20, validation error 46.400000 %
epoch 9, minibatch 20/20, validation error 46.400000 %
epoch 10, minibatch 20/20, validation error 46.400000 %
training @ iter =  200
epoch 11, minibatch 20/20, validation error 46.400000 %
epoch 12, minibatch 20/20, validation error 46.400000 %
epoch 13, minibatch 20/20, val

The training process for function test_mlp_parity ran for 0.45m


In [17]:
accuracy = test_mlp_parity(learning_rate=0.1, L1_reg=0.00, L2_reg=0.00, n_epochs=1000,
             batch_size=80, n_hidden=200, n_hiddenLayers=5,
             verbose=True, n_bit = 12)

... building the model
... training
training @ iter =  0
epoch 1, minibatch 12/12, validation error 49.166667 %
     epoch 1, minibatch 12/12, test error of best model 42.500000 %
epoch 2, minibatch 12/12, validation error 50.833333 %
epoch 3, minibatch 12/12, validation error 50.833333 %
epoch 4, minibatch 12/12, validation error 50.416667 %
epoch 5, minibatch 12/12, validation error 50.208333 %
epoch 6, minibatch 12/12, validation error 49.583333 %
epoch 7, minibatch 12/12, validation error 49.791667 %
epoch 8, minibatch 12/12, validation error 49.375000 %
training @ iter =  100
epoch 9, minibatch 12/12, validation error 48.958333 %
     epoch 9, minibatch 12/12, test error of best model 47.500000 %
epoch 10, minibatch 12/12, validation error 47.916667 %
     epoch 10, minibatch 12/12, test error of best model 47.500000 %
epoch 11, minibatch 12/12, validation error 47.916667 %
epoch 12, minibatch 12/12, validation error 47.708333 %
     epoch 12, minibatch 12/12, test error of best m

The training process for function test_mlp_parity ran for 0.42m


#### Bullet 3: RNN

In [4]:
accuracy = test_rnn_parity(learning_rate=1, L1_reg=0.00, L2_reg=0.1, n_epochs=30,
             batch_size=8, n_hidden=20, verbose=False, n_bit = 8)

... building the model
... training
Optimization complete.
Best validation score of 51.612903 % obtained at iteration 125, with test performance 44.791667 %


The training process for function test_rnn_parity ran for 0.59m


In [3]:
accuracy = test_rnn_parity(learning_rate=0.1, L1_reg=0.00, L2_reg=0.00, n_epochs=30,
             batch_size=20, n_hidden=100, verbose=False, n_bit = 12)

... building the model
... training
Optimization complete.
Best validation score of 51.600000 % obtained at iteration 50, with test performance 46.000000 %


The training process for function test_rnn_parity ran for 0.50m


Both a 2-layer MLP and a deep MLP can learn parity function and reaches a accuracy of 100%. But temporarily I cannot find parameteres for RNN to learn the parity function to reach 100%.